src/hotspot/share/cds/heapShared.cpp

```diff
@@ -343,21 +346,110 @@ oop HeapShared::archive_object(oop obj) {
   return archived_oop;
 }
 
-void HeapShared::archive_klass_objects() {
+class KlassToOopHandleTable: public ResourceHashtable<Klass*, OopHandle,
+    36137, // prime number
+    AnyObj::C_HEAP,
+    mtClassShared> {
+public:
+  oop get_oop(Klass* k) {
+    MutexLocker ml(ScratchObjects_lock, Mutex::_no_safepoint_check_flag);
+    OopHandle* handle = get(k);
+    if (handle != NULL) {
+      return handle->resolve();
+    } else {
+      return NULL;
+    }
+  }
+  void set_oop(Klass* k, oop o) {
+    MutexLocker ml(ScratchObjects_lock, Mutex::_no_safepoint_check_flag);
+    OopHandle handle(Universe::vm_global(), o);
+    bool is_new = put(k, handle);
+    assert(is_new, "cannot set twice");
+  }
+  void remove_oop(Klass* k) {
+    MutexLocker ml(ScratchObjects_lock, Mutex::_no_safepoint_check_flag);
+    OopHandle* handle = get(k);
+    if (handle != NULL) {
+      handle->release(Universe::vm_global());
+      remove(k);
+    }
+  }
+};
+
+static KlassToOopHandleTable* _scratch_java_mirror_table = NULL;
```

Comment:
Any reason why `_scratch_basic_type_mirrors` is a static member of `HeapShared` but `_scratch_java_mirror_table` is not?

Reply:
I moved this field as a static member of `HeapShared` as well.